Snap to integer zoom level

quinncnl opened the pull request:
Add an option to snap to integer zoom level: https://github.com/maplibre/maplibre-gl-js/issues/1591

Example: https://tracestrack.s3-website.nl-ams.scw.cloud/dist/
Example: https://tracestrack.s3-website.nl-ams.scw.cloud/dist/#12.40/11.7933/53.4402
jsfiddle: https://jsfiddle.net/142rftnx/1/

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented:
Thanks for taking the time to contribute!
Quick question, What about touch gestures?
Also not that there's a similar behavior to snap bearing to north, might be worth getting inspiration from, maybe...

quinncnl commented:
Thanks for the note! I'll check those :)

quinncnl commented:
I've done some changes to `ScrollZoom`. For `TagDragZoom` and `TwoFingersTouchZoom` it requires more investigation and probably a bigger change. Can we ignore the touch zoom for now?

HarelM commented:
I don't think we can ignore the touch events without a flag that is specifying to which gestures the interger zoom applies I guess.
Also I'm not sure the animation of the zoom scroll in the demo is working well...
Can you add hash control to the demo so that zoom level will be easy to see?

quinncnl commented:
OK will add hash control. For the touch event it'll take some time since it's quite complicated.

HarelM commented:
What I meant to say is that if you decide not to implement touch handling the flag specifying the integer zoom should be something like:
snapToIntegerZoom: {touch: false, scroll: true, controls: true}
And not a simple boolean, this will on one hand allow more granularity and also provide a way to implement just the relevant part.
When all are implemented, true and false can be added to indicate "apply to all" kind of configuration.

HarelM commented:
Any updates on this?

quinncnl commented:
Hey @HarelM sorry I got busy. Will try to followup on it this weekend. I'll break it into smaller pieces as you suggested.

quinncnl commented:
Hi @HarelM I've updated the example map and added a jsfiddle. Could you please check it and provide feedback? Once it looks good I'll fix the tests and docs.

HarelM commented on src/ui/map.ts:
```diff
@@ -1019,6 +1027,8 @@ export class Map extends Camera {
      */
     getMaxZoom(): number { return this.transform.maxZoom; }
 
+    getSnapToIntegerZoomOptions(): SnapToIntegerZoomOptions { return this._snapToIntegerZoomOptions; }
```
Consider getting a string of the relevant option instead of returning the object, this way in the future you can check if the value of this is a boolean and return that value, obfuscating the internal implementation.

quinncnl replied:
Could you elaborate what kind of string? When we move to single boolean toggle we can still use this options object right?

HarelM replied:
Use `type SnapType = "box" | "scroll" ...` as the input parameter to this method.

quinncnl replied:
Done

HarelM commented on src/ui/map.ts:
```diff
@@ -1008,6 +1014,8 @@ export class Map extends Camera {
         } else throw new Error('maxZoom must be greater than the current minZoom');
     }
 
+    setSnapToIntegerZoom(snap: SnapToIntegerZoomOptions): Map { this._snapToIntegerZoomOptions = snap; return this; }
```
Use multiple line instead of a single one.

quinncnl replied:
Done

HarelM commented on src/ui/handler/tap_zoom.ts:
```diff
@@ -57,7 +57,7 @@ export class TapZoomHandler implements Handler {
                 cameraAnimation: (map: Map) => map.easeTo({
                     duration: 300,
-                    zoom: tr.zoom + 1,
+                    zoom: Math.round(tr.zoom + 1),
```
This should be using the check?

quinncnl replied:
Done

HarelM commented on src/ui/handler/tap_zoom.ts:
```diff
@@ -68,7 +68,7 @@ export class TapZoomHandler implements Handler {
                 cameraAnimation: (map: Map) => map.easeTo({
                     duration: 300,
-                    zoom: tr.zoom - 1,
+                    zoom: Math.round(tr.zoom - 1),
```
This should be using the check?

quinncnl replied:
Done

HarelM commented:
I think this direction is OK.

sbachinin commented:
Noticed that in your [jsfiddle](https://jsfiddle.net/L2zdp3eq/1/) if you zoom out to 3.0, you can't zoom in back

HarelM commented on src/ui/map.ts:
```diff
@@ -61,6 +61,8 @@ import type {QueryRenderedFeaturesOptions, QuerySourceFeatureOptions} from '../s
 
 const version = packageJSON.version;
 
+export type SnapToIntegerZoomType = 'boxZoom' | 'scrollZoom' | 'clickZoom' | 'tapZoom';
```
I think you can use `keyof SnapToIntegerZoomOptions`...?

quinncnl replied:
Nice! I'm not expert of typescript so your help is useful :)

quinncnl replied:
Fixed

HarelM commented on src/ui/map.ts:
```diff
@@ -1019,6 +1032,19 @@ export class Map extends Camera {
      */
     getMaxZoom(): number { return this.transform.maxZoom; }
 
+    shouldSnapToIntegerZoom(type: SnapToIntegerZoomType): boolean {
```
`SnapToIntegerZoomType` contains 4 values, this switch case contains 3, is it intentional?

qingcai-tomtom replied:
I missed it! Thanks

quinncnl replied:
Fixed

HarelM replied:
Since SnapToIntegerZoomType is keyof SnapToIntegerZoomOptions I think you can simply use return `this._snapToIntegerZoomOptions[type]` can't you?

quinncnl replied:
Yes indeed!

quinncnl commented on src/ui/camera.ts:
```diff
@@ -113,6 +138,11 @@ export type CameraForBoundsOptions = CameraOptions & {
+    /**
+     * Whether the map should snap to integer zoom levels during the transition.
+     */
+    shouldSnapToIntegerZoom?: boolean;
```
Note: this is for BoxZoom only.

HarelM commented on test/build/min.test.ts:
```diff
@@ -29,7 +29,7 @@ describe('test min build', () => {
         // Need to be very frugal when it comes to minified script.
         // Most changes should increase less than 1k.
-        const increaseQuota = 1024;
+        const increaseQuota = 1600;
```
Better to increase the expectedBytes instead of the quota.

quinncnl replied:
Indeed. Updated

HarelM commented on src/ui/map.ts:
```diff
@@ -1036,6 +1031,35 @@ export class Map extends Camera {
+        if (options === null || options === undefined) {
+            this._snapToIntegerZoomOptions = {boxZoom: false, clickZoom: false, scrollZoom: false, tapZoom: false};
+        } else {
+            this._snapToIntegerZoomOptions = options;
```
I think this will reset other snap options, wouldn't it? i.e. if I set scroll to true in the first call and set box to true in the second call, I'll reset scroll configuration.
Consider doing a for loop here maybe? Not sure if it's a great idea though...

HarelM replied:
Or maybe consider the API to be similar to the get/should

quinncnl replied:
I see your point. I think replacing the entire options is more logical here because we are setting `SnapToIntegerZoomOptions` instead of `type: SnapToIntegerZoomType`. It's also same like other setters like `style: StyleSpecification`.

> Or maybe consider the API to be similar to the get/should

Something like `setSnapToIntegerZoom(type: SnapToIntegerZoomType, value: boolean)`? Normally this options is set once at the map init. Using this kind of setter is not very convenient in this common case. As you might find codes like:
```
setSnapToIntegerZoom("boxScroll", true);
setSnapToIntegerZoom("clickScroll", true);
```
instead of simple setOptions({boxScrolll: true, clickScroll: true})

HarelM commented on src/ui/handler/scroll_zoom.ts:
```diff
@@ -193,6 +197,24 @@ export class ScrollZoomHandler implements Handler {
+
+            zoomTarget = Math.round(zoomTarget);
+
+            this._map.easeTo({
```
It seems like there is a explicit animation code for regular zoom, how come this can be avoided using easeTo instead of writing the animation code (using frames etc)?

quinncnl replied:
The explicit animation is about smoothOutEasing which doesn't have a targetZoom. The end zoom is based on an easing function which is hard to ease to an integer level. So I just skip that and use a simple `easeTo` instead.

HarelM commented:
@quinncnl did you check/fix the issue reported by @sbachinin?

quinncnl commented:
> @quinncnl did you check/fix the issue reported by @sbachinin?

I can't seem to reproduce the scenario. @sbachinin could you try again? Maybe it was before my refator to scrollZoom.

sbachinin commented:
This particular [jsfiddle](https://jsfiddle.net/L2zdp3eq/1/) shows the same bad behaviour.
It could be broken because of using some outdated bundle, I don't know.

Here's what I get when I try to zoom in:

https://github.com/maplibre/maplibre-gl-js/assets/11789697/c8fd5837-7b4a-4fcd-87ea-03a7c8e044c7

quinncnl commented:
Hi @sbachinin I updated the demo. Could you try again? It seems we have to round the zoom at the end of inertia animation. If this is not acceptable can we have a separate toggle for it, like `trackpadScroll` and do it later? @HarelM

HarelM commented:
I would expect the last commit to be added with a test to make sure this is the expected behavior.

sbachinin commented:
@quinncnl, all the demos given here still demonstrate zooming problems for me. (Chrome 122.0.6261.129)
When I roll mousewheel forward:
first several "clicks" are ignored, then I'm jumping to zoom 3.
When I roll mousewheel backward:
first several "clicks" are ignored, then begins a normal zoom out.

quinncnl commented:
I tried a different computer on Windows and there it jumps zoom by 2 with single scroll. Looks like I need to test with different browser/OS/scroll device(trackpad/mouse) as they have different scroll values.

quinncnl commented:
Hi @HarelM , do you think it's okay to move on without scroll_zoom? At least then people can build upon it.

HarelM commented:
There are missing tests and reported errors. How do you suggest to address them?

quinncnl commented:
> There are missing tests and reported errors. How do you suggest to address them?

I will revert all changes related to scrollZoom as it's hard to fix with my knowledge. It does not pre-calculate a targetZoom making it hard to achieve integer final zoom.

I'll add some tests to supported handlers.

HarelM commented:
@quinncnl are you still working on this?
Seems like this needs more work than you originally signed up for, and I think the implementation is too lacking to be considered as a full feature.
I'm trying to clean up stale PRs.
Let me know if I can close this one.

quinncnl commented:
Hi @HarelM indeed it requires more work still. I'm quite occupied and like to park it. If someone can pick it up that would nice!

HarelM commented:
I don't think there is ATM.
Closing this for now, if anyone would like to work on it in the future they would be able to find it here.
